gradle/lib/dependabot/gradle/utils/auth_details_finder.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module Gradle
+    module Utils
+      class AuthDetailsFinder
+
+        def initialize(credentials)
+          @credentials = credentials
+        end
+
+        def auth_details(maven_repo_url)
+          cred =
+            credentials.select { |c| c["type"] == "maven_repository" }.
+            find do |c|
+              cred_url = c.fetch("url").gsub(%r{/+$}, "")
+              next false unless cred_url == maven_repo_url
+
+              c.fetch("username", nil)
+            end
+
+          return gitlab_auth_details(maven_repo_url) unless cred
```

Comment:
The code might be unclear, the first thing done in the gitlab_auth_details method is to check if the maven repository path matches a gitlab maven repository, otherwise `{}` is returned.

My first approach actually was to add an option to provide custom headers when defining maven_repositories, but a change would still be needed to allow the header being added to the request, as the current solution defaults to basic auth.